libvcx/src/api_vcx/api_handle/connection.rs

```diff
@@ -42,62 +118,143 @@ where
 // ----------------------------- CONSTRUCTORS ------------------------------------
 pub async fn create_inviter(pw_info: Option<PairwiseInfo>) -> LibvcxResult<u32> {
     trace!("create_inviter >>>");
-    store_connection(Connection::create_inviter(&get_main_profile()?, pw_info).await?)
+    let profile = get_main_profile()?;
+
+    let pw_info = pw_info.unwrap_or(PairwiseInfo::create(&profile.inject_wallet()).await?);
+    let con = InviterConnection::new_inviter("".to_owned(), pw_info);
+
+    add_connection(con)
 }
 
-pub async fn create_invitee(invitation: &str) -> LibvcxResult<u32> {
+pub async fn create_invitee(_invitation: &str) -> LibvcxResult<u32> {
     trace!("create_invitee >>>");
+
     let profile = get_main_profile()?;
-    store_connection(
-        Connection::create_invitee(&profile, into_did_doc(&profile, &deserialize(invitation)?).await?).await?,
-    )
+    let pairwise_info = PairwiseInfo::create(&profile.inject_wallet()).await?;
+
+    let con = InviteeConnection::new_invitee("".to_owned(), pairwise_info);
+
+    add_connection(con)
 }
 
 // ----------------------------- GETTERS ------------------------------------
 pub fn get_thread_id(handle: u32) -> LibvcxResult<String> {
     trace!("get_thread_id >>> handle: {}", handle);
-    CONNECTION_MAP.get(handle, |connection| Ok(connection.get_thread_id()))
+
+    let lock = CONNECTION_MAP.read()?;
+    let con = lock.get(&handle).ok_or_else(|| {
```

Comment:
there's a bit of code duplication here. Could it be moved to a function similar to `get_cloned_connection`? something like:

```
fn get_borrowed_connection_from_lock<'a>(lock: &'a RwLockReadGuard<HashMap<u32, GenericConnection>>, handle: u32) -> LibvcxResult<&'a GenericConnection>
```

Reply:
Not sure what the benefit would really be. 

You'd still have to call `CONNECTION_MAP.read()` to get the lock guard, which must live for the entire duration of the operation, so we essentially only get rid of the `lock.get()` and replace it with `get_borrowed_connection_from_lock()`.

I don't really like the duplication either but in terms of ownership it has to behave like that. 
I'll look into using a function that takes a closure, or maybe a macro taking a token tree.

Reply:
kk. just to be clear, was referring to the `lock.get` and the mapping to the same error type, as it's duplicated 6 times;

```
fn get_borrowed_connection_from_lock<'a>(lock: &'a RwLockReadGuard<HashMap<u32, GenericConnection>>, handle: u32) -> LibvcxResult<&'a GenericConnection> {
    lock.get(&handle).ok_or_else(|| LibvcxError::from_msg(
        LibvcxErrorKind::ObjectAccessError,
        format!("No connection found for handle: {}", handle),
    ))
}
```

Reply:
Ah, fair. I was speaking from memory and forgot that there's also the common error being returned if a connection is not found for a handle, apart from the attribute specific error returned if the connection is in an invalid state and does not contain the attribute.

Nevertheless, I added `get_con_attribute_with_closure` to factor out the duplicate code.